File: vmim/src/main/java/com/vmloft/develop/library/im/bean/IMConversation.java
package com.vmloft.develop.library.im.bean;

import com.hyphenate.chat.EMConversation;
import com.hyphenate.chat.EMMessage;

import com.vmloft.develop.library.im.utils.IMChatUtils;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * Create by lzan13 on 2019/5/21 15:00
 *
 * 自定义 IMConversation 会话对象
 */
public class IMConversation implements Serializable {

    // 内部会话对象
    private EMConversation mConversation;

    // 会话 id
    private String mId;
    // 草稿
    private String mDraft;
    // 是否置顶
    private boolean isTop;
    // 未读状态
    private boolean isUnread;
    // 最后一次更新时间
    private long mLastTime;
    // 最后一条消息
    private IMMessage mLastMessage;

    /**
     * 空构造
     */
    public IMConversation() {
    }

    /**
     * 根据传入内部需要的会话对象构造 IMConversation
     */
    public IMConversation(EMConversation conversation) {
        mConversation = conversation;
    }

    public void setConversation(EMConversation mConversation) {
        this.mConversation = mConversation;
    }

    public String getId() {
        mId = mConversation.conversationId();
        return mId;
    }

    public String getDraft() {
        mDraft = IMChatUtils.getConversationDraft(mConversation);
        return mDraft;
    }

    public void setDraft(String draft) {
        this.mDraft = draft;
        IMChatUtils.setConversationDraft(mConversation, draft);
    }

    public boolean isTop() {
        isTop = IMChatUtils.getConversationTop(mConversation);
        return isTop;
    }

    public void setTop(boolean top) {
        isTop = top;
        IMChatUtils.setConversationTop(mConversation, top);
    }

    public boolean isUnread() {
        isUnread = IMChatUtils.getConversationUnread(mConversation);
        return isUnread;
    }

    public void setUnread(boolean unread) {
        isUnread = unread;
        IMChatUtils.setConversationUnread(mConversation, unread);
    }

    public long getLastTime() {
        mLastTime = IMChatUtils.getConversationLastTime(mConversation);
        return mLastTime;
    }

    public void setLastTime(long lastTime) {
        this.mLastTime = lastTime;
        IMChatUtils.setConversationLastTime(mConversation, lastTime);
    }

    public IMMessage getLastMessage() {
        if (mConversation.getAllMsgCount() == 0) {
            mLastMessage = new IMMessage();
        }else{
            mLastMessage = new IMMessage(mConversation.getLastMessage());
        }
        return mLastMessage;
    }

    public void setLastMessage(IMMessage mLastMessage) {
        this.mLastMessage = mLastMessage;
    }

    /**
     * 获取会话的所有消息
     */
    public List<IMMessage> getAllMessage() {
        List<IMMessage> result = new ArrayList<>();
        List<EMMessage> list = mConversation.getAllMessages();
        for (EMMessage message : list) {
            result.add(new IMMessage(message));
        }
        return result;
    }

    /**
     * 加载更多消息
     *
     * @param msgId 开始加载的消息 Id
     * @param limit 加载条数
     */
    public List<IMMessage> loadMoreMessage(String msgId, int limit) {
        List<IMMessage> result = new ArrayList<>();
        List<EMMessage> list = mConversation.loadMoreMsgFromDB(msgId, limit);
        for (EMMessage message : list) {
            result.add(new IMMessage(message));
        }

        return result;
    }

    /**
     * 当会话 id 相同就认为两个会话相同
     */
    @Override
    public boolean equals(Object o) {
        if (o instanceof IMConversation) {
            IMConversation bean = (IMConversation) o;
            return mId.equals(bean.mId);
        }
        return super.equals(o);
    }

}

File: vmim/src/main/java/com/vmloft/develop/library/im/bean/IMConversation.kt
package com.vmloft.develop.library.im.bean

import android.os.Parcelable
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey

import kotlinx.parcelize.Parcelize

/**
 * Create by lzan13 on 2022/8/10 22:00
 * 描述：会话 bean
 */
@Entity(tableName = "conversation")
@Parcelize
data class IMConversation(
    @PrimaryKey
    var chatId: String = "", // 会话Id
    var type: Int = 0, // 类型 0-单聊 1-群聊 2-聊天室
    var content: String = "", // 内容
    var time: Long = 0, // 时间戳
    var top: Boolean = false, // 置顶
    var receiveCount: Int = 0, // 接收数量
    var sendCount: Int = 0, // 发送数量
    var unread: Int = 0, // 未读

    var draft: String = "", // 草稿
    @Ignore
    var msgList: MutableList<IMMessage> = mutableListOf(), // 消息列表
) : Parcelable {

    /**
     * 加载更多消息
     *
     * @param msgId 开始加载的消息 Id
     * @param limit 加载条数
     */
    fun loadMoreMsg(msgId: String?, limit: Int): List<IMMessage> {
        val result = mutableListOf<IMMessage>()
//        for (message in list) {
//        }
        return result
    }

    /**
     * 添加消息
     */
    fun addMessage(message: IMMessage) {
        if (msgList.contains(message)) {
            msgList.remove(message)
        }
        msgList.add(message)
    }

    /**
     * 添加消息
     */
    fun addMessages(messages: List<IMMessage>, history: Boolean = true) {
        if (history) {
            msgList.addAll(0, messages)
        } else {
            msgList.addAll(messages)
        }
    }

    /**
     * 移除消息
     */
    fun removeMessage(message: IMMessage) {
        msgList.remove(message)
    }

    /**
     * 更新消息
     */
    fun updateMessage(message: IMMessage) {
    }

    /**
     * 清空消息
     */
    fun clearMsg() {
        msgList.clear()
    }

    /**
     * 消息数量增加
     */
    fun msgCountAdd(isSend: Boolean = false) {
        if (isSend) {
            sendCount++
        } else {
            receiveCount++
        }
    }

    /**
     * 当会话 id 相同就认为两个会话相同
     */
    override fun equals(other: Any?): Boolean {
        if (other is IMConversation) {
            return chatId == other.chatId
        }
        return super.equals(other)
    }
}